File: Course 1 - Algorithm Toolbox/week4_divide_and_conquer/2_majority_element/majority_element.cpp
#include <algorithm>
#include <iostream>
#include <vector>
#include <unordered_map>

using std::vector;


int get_majority_element(vector<int> &a, int left, int right) {
    std::unordered_map<int, int> u = {    };
    float arr_len_div2 = float(right) / 2;
    for (int i = 0; i < right; i++)
    {
        int item = a[i];
        std::unordered_map<int, int>::const_iterator got = u.find(item);
        if (got == u.end())
            u[item] = 1;
        else {
            u[item] = u[item] + 1;
            if (u[item] > arr_len_div2)
            {
                return 1;
            }
        }
    }
    return -1;
}

int main() {
  int n;
  std::cin >> n;
  vector<int> a(n);
  for (size_t i = 0; i < a.size(); ++i) {
    std::cin >> a[i];
  }
  std::cout << (get_majority_element(a, 0, a.size()) != -1) << '\n';
}

File: Course 1 - Algorithm Toolbox/week4_divide_and_conquer/2_majority_element/majority_element.py
# Uses python3
import sys

def get_majority_element(a, left, right):
    if left == right:
        return -1
    if left + 1 == right:
        return a[left]
    #write your code here
    return -1

def majority_element(arr,left,right):
    my_dict = {}
    arr_len_div2 = right/2
    for item in arr:
        if item in my_dict:
            my_dict[item] += 1
            if my_dict[item] > arr_len_div2:
                return 1
        else:
            my_dict[item] = 1
    return -1

if __name__ == '__main__':
    input = sys.stdin.read()
    n, *a = list(map(int, input.split()))
    if majority_element(a, 0, n) != -1:
        print(1)
    else:
        print(0)
